### ISSUE TITLE
Format date columns

### ISSUE THREAD

**Issue by rdstern** [IDEMSInternational/R-Instat#4282]
We either need a dialogue to be able to format date columns.  Or it could be a 4th button on the Make Date dialogue.  It should be easy to add?

**Comment by Patowhiz** [comment]
@rdstern should this functionality produce a new date column or just modify the existing one with the format selected?

**Comment by rdstern** [comment]
I suggest this simply modifies the current Date column.

**Comment by dannyparsons** [comment]
It's more "Use Date" than "Make Date" since it requires a Date column and it returns a character column.
Note that formatting a Date column returns a character column so it's a one way process - in general you can't reconstruct the Date after formatting. So in this way it's more like the options on the Use Date dialog. Since it requires a format string like "%a %b %d" I can see why a separate dialog might be useful.

**Comment by rdstern** [comment]
I accept that it is better placed on the Use Date dialogue.  I am not sure it is worth a separate dialogue as (as Danny says) it becomes a character column.  So it is more for cosmetic purposes.

Danny has also suggested that there could perhaps be a special keyboard to help users to prepare the format they would like.  Then it is simply the command format(date_column,"string") where (for example the string could be **"%d %b %Y"** to give say **23 May 2012**, or "%d %b" would be just **23 May**.

Here is my first go at the keyboard - which would have tool-tips to add explanations.  There could be 5 columns with headings as follows:

Day/Week   Weekday  Month  Year  Separator
%d                %u             %m      %y    <space>
%j                 %w             %b      %Y          -  (That's a dash - minus sign)
_____                                           _____
%U                %a             %B       %C         /
%W               %A                                      _  (That's an underscore)

(Try editing this issue to see the layout better!)

**Comment by dannyparsons** [comment]
I've tidied up the table a bit :)

It looks good - I like the headers. I just swapped %y and %Y around because I think %Y is the most common one.

**Comment by lilyclements** [comment]
I think this could be revisited and still possible. In "Make Date", we have this

![Image](https://github.com/user-attachments/assets/4c432645-2b41-4d0a-9949-74d95fc30a58)

How about we have something like this, but in reverse, on the "Use Date" dialog? So we have a "Specify Format" option again, and you can select a string of what you'd like to show, or write it in too optionally. 

@rdstern what do you think?

**Comment by lilyclements** [comment]
@rdstern great! This should be straightforward as you say. I did the R code for it the other day, except I was getting a little held up on using it with shifted days. But that can be disabled until that bit of R is implemented. We now have the functionality for having these different formatted dates, which I’ll give the R code to when I am next on my laptop.

**Comment by rdstern** [comment]
Now to the section of the date keyboard.  The section is called Format

I suggest 4 rows with the top row giving complete formats and the others giving the abbreviations.  All with tooltips

Row 1: 4 Jun 25, 4 Jun 2025, Jun 4, Jun 04, %H
Row 2: %d, %e, %j, %U, %I
Row 3: %m, %b, %B, %P, %M
Row 4: %Y, %y, %a, %A, %S

The codes and tooltips, when different to the keys are as follows:
Row 1
4 Jun 25 gives %d 

To be continued

**Comment by lilyclements** [comment]
@rdstern great. So are you suggesting something like this (with the row format discussion from your previous message being in the "Format Description" bit)

![Image](https://github.com/user-attachments/assets/d99cc477-f24a-4e07-93a0-4829c1326f70)

The R code here is:

```r



```


I could not get `DescTools` to work in this case. After a google, it said: 

Use `format()` directly for date-specific formatting like day-of-year

```r
format(x$date, "%j")
```

Only use `DescTools::Format()` when:

* You need consistent numeric formatting (digits, perc, etc.),
* Or you're formatting mixed-type vectors for output.

So I think in our case we can use just `format`

**Comment by rdstern** [comment]
@lilyclements I used the DescTools::Format(  , fmt=(  , "  ") a lot for the dates when designing the keyboard and was pretty impressed!  

See their documentation on page 230 of their manuel.  I found also you could use any of the codes like %F, etc from format(  ).  So I didn't see any problem, and for dates it was nice and simple.  
In the calculator I added the time equivalents, for hours, minutes, seconds

**Comment by lilyclements** [comment]
@rdstern great - thanks for outlining that. I've now added that into the most recent version of `databook` for this function. So we are now using `DescTools` not the base format.

**Comment by lilyclements** [comment]
Yes, that should all work. Hopefully the R side is all good and this can be looked at while I'm away if anyone is short of a task

**Comment by lilyclements** [comment]
@ARMSTRONGOPONDO great - this is perfect. You want to remove the bit circled in red, as you correctly identified, when "Format" is checked.
Then we want Format Description to be similar to the dlgMakeDate dialog under the "Single Date" tab (with this outlined above in this issue)
And we want "New Format" to only be visible on the "Format Description" tab (which you might well have set up already, but just saying in case that wasn't clear)

**Comment by berylwaswa** [comment]
@ARMSTRONGOPONDO did you ever finish working on this?

**Comment by ARMSTRONGOPONDO** [comment]
@berylwaswa the PR no: #9866 was the one trying to handle this, I just have to confirm, if it that PR has captured everything or not before this issue is closed.

**Comment by berylwaswa** [comment]
@ARMSTRONGOPONDO please do so

**Comment by ARMSTRONGOPONDO** [comment]
All the outlined issues from this discussion seem to have been resolved closing this issue now,
